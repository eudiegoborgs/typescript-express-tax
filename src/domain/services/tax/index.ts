export interface TaxServiceInterface {
  calculate(amount: number, tax: number): string;
}

export class TaxService implements TaxServiceInterface {
  calculate(amount: number, tax: number): string {
    const total = (1 + tax / 100) * amount;
    return total.toFixed(2);
  }
}