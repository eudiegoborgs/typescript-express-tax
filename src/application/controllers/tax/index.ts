import Joi from 'joi';
import { TaxServiceInterface } from "../../../domain/services/tax";

export class TaxController {
  constructor(private readonly service: TaxServiceInterface) {}

  calculate(req, res): void {
    const schema = Joi.object({
      tax: Joi.number().required(),
      amount: Joi.number().required(),
    })
    const {tax, amount} = req.body;
    const result = schema.validate({tax, amount});
    if (result.error) {
      throw {code: 403, message: result.error}
    }
    res.json({
      valueWithTaxes: this.service.calculate(amount, tax),
    })
  }
}