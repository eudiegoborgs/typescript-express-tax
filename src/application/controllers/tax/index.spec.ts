import { TaxController } from "./";
import { TaxServiceInterface } from "../../../domain/services/tax";

export class TaxServiceMock implements TaxServiceInterface {
  calculate(amount: number, tax: number): string {
    return "10";
  }
}

const service = new TaxServiceMock();

const controller = new TaxController(service);

describe("Testing TaxController class", () => {
  it("Testing calculate function", () => {
    const req = {
      body: {
        tax: 2.2,
        amount: 10000
      }
    }

    const res = {
      json: data => {
        expect(data).toEqual({valueWithTaxes: "10"});
      }
    }

    const spy = jest.spyOn(res, 'json');

    controller.calculate(req, res);

    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy).toHaveBeenCalledWith(
      {valueWithTaxes: "10"}
    );
  });

  it("Testing calculate function without data", () => {
    const req = {
      body: {}
    }

    const res = {}

    expect(() => controller.calculate(req, res)).toThrowErrorMatchingSnapshot();
  });

  it("Testing calculate function without amount", () => {
    const req = {
      body: {
        tax: 2.2,
      }
    }

    const res = {}

    expect(() => controller.calculate(req, res)).toThrowErrorMatchingSnapshot();
  });

  it("Testing calculate function without amount", () => {
    const req = {
      body: {
        amount: 10000,
      }
    }

    const res = {}

    expect(() => controller.calculate(req, res)).toThrowErrorMatchingSnapshot();
  });

  it("Testing calculate function with amount", () => {
    const req = {
      body: {
        amount: 10000,
      }
    }

    const res = {}

    expect(() => controller.calculate(req, res)).toThrowErrorMatchingSnapshot();
  });
})