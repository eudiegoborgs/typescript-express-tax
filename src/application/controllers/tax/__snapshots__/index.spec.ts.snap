// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing TaxController class Testing calculate function without amount 1`] = `[ValidationError: "amount" is required]`;

exports[`Testing TaxController class Testing calculate function without amount 2`] = `[ValidationError: "tax" is required]`;

exports[`Testing TaxController class Testing calculate function without data 1`] = `[ValidationError: "tax" is required]`;
